### ISSUE TITLE
How do I create a pygfx geometry with texture coordinates from an open3d triangle mesh?

### ISSUE THREAD

**Issue by cansik** [pygfx/pygfx#641]
I use [open3d](https://github.com/isl-org/Open3D) a lot to manipulate 3d models and point clouds. Open3d's built-in 3d engine is ok for a lot of things, but pygfx is much easier to use. That's why I like to use pygfx to render the 3d models, and to efficiently transfer the open3d [triangle mesh](https://www.open3d.org/docs/release/python_api/open3d.geometry.TriangleMesh.html) to pygfx geometry, I've implemented the following method. The method is based on [pygfx.geometries.geometry_from_trimesh](https://pygfx.readthedocs.io/en/latest/_autosummary/geometry/pygfx.geometries.geometry_from_trimesh.html).

```python
def load_from_open3d(path: str) -> gfx.Geometry:
    # load open3d mesh as model and extract first mesh
    o3d_model: rendering.TriangleMeshModel = io.read_triangle_model(path)
    o3d_mesh_info: rendering.TriangleMeshModel.MeshInfo = o3d_model.meshes[0]
    o3d_mesh: geometry.TriangleMesh = o3d_mesh_info.mesh

    # create geometry
    triangle_uvs = np.array(o3d_mesh.triangle_uvs, dtype=np.float32)
    triangles = np.array(o3d_mesh.triangles, dtype=np.uint32)

    vertex_normals = np.array(o3d_mesh.vertex_normals, dtype=np.float32)
    vertex_colors = np.array(o3d_mesh.vertex_colors, dtype=np.float32)
    vertices = np.array(o3d_mesh.vertices, dtype=np.float32)

    # translate uv coordinates from opengl to wgpu (uv.y = 1 - uv.y)
    triangle_uvs_wgpu = (triangle_uvs * np.array([1, -1]) + np.array([0, 1])).astype(np.float32)

    # create pygfx geometry
    gfx_geometry = gfx.Geometry(indices=triangles, positions=vertices,
                                normals=vertex_normals, texcoords=triangle_uvs_wgpu)

    return gfx_geometry
```

Take a look at the Minimal Reproducible Example (MRE): [cansik/gfx-open3d-example](https://github.com/cansik/gfx-open3d-example)

#### Problem Description
Loading the geometry works, however, when adding a texture to the model, the texture coordinates are completely off (see Figure 1).

<img width="720" alt="image" src="https://github.com/pygfx/pygfx/assets/5220162/8b683901-b2bc-4b17-a597-a83bb3b87e2a">

*Figure 1: Left is loaded by trimesh, right side is loaded by open3d*

#### Investigation
The `texcoords` shapes of the two models are different:

- loaded by trimesh = `(24, 2)`
- loaded by open3d = `(36, 2)`

Which is also different for trimesh, if the `load()` method is used with `process=Flase, maintain_order=True` (see Figure 2):

```python
t_mesh: Trimesh = trimesh.load(path, process=False, maintain_order=True)
# shape: (8, 2)
```

<img width="723" alt="image" src="https://github.com/pygfx/pygfx/assets/5220162/cc79a328-1c5e-4024-8175-3613feae11a4">

*Figure 2: Left is loaded by trimesh (without processing), right side is loaded by open3d*

Which led me to the following issues in the open3d repository related to obj mesh loading: https://github.com/isl-org/Open3D/issues/2614, https://github.com/isl-org/Open3D/issues/2978, https://github.com/isl-org/Open3D/issues/6137

Enabling or disabling `enable_post_processing` does not help for open3d. And I know this is a problem on the open3d side. However, the mesh can be rendered with open3d, so it should be possible in other engines somehow. 

#### Question
It would be great to know if it is somehow possible to load the model (and textcoords) in pygfx without first saving the mesh as an obj and reloading it with trimesh?

**Comment by almarklein** [comment]
My guess is that the `triangle_uvs` are the uv's per index, not per vertex. That explains both their shape and name 😄 

With these lines you can convert them to `vertex_uvs`:
```py
    vertex_uvs = np.zeros((len(vertices), 2), np.float32)
    vertex_uvs[triangles.flat] = triangle_uvs
```

I think the solution is in this direction. Although it seems like it's not quite there yet:
<img width="379" alt="image" src="https://github.com/pygfx/pygfx/assets/3015475/d0477c75-ff23-44e2-b66d-83eae23d2cc2">


Is the trimesh version a good reference, or could that be wrong as well? Probably best to compare it with how it looks in open3d?

**Comment by cansik** [comment]
Oh, thanks for that simple solution ✨ I guess the wgpu conversion has not been applied in your rendering. It's seems ok now:

<img width="449" alt="image" src="https://github.com/pygfx/pygfx/assets/5220162/6cca2289-3a45-4040-a422-fb9ce1e25f32">

**Comment by cansik** [comment]
@almarklein Would you be open to a PR that adds a compatibility method to load open3d geometries and materials like the one for trimesh? Or do you want to keep pygfx clean and prefer an additional package / library?

**Comment by almarklein** [comment]
I think a `_compat.geometry_from_open3d` is very welcome. @Korijn thoughts? I think since these compatibility functions are relatively small, there's no harm in adding them.

**Comment by Korijn** [comment]
Yes please! It's useful to centrally document these integrations. :) People can extend them in the future as well.
